collector/netclass_rtnl_linux.go

```diff
@@ -0,0 +1,489 @@
+		pushStats(ch, c.subsystem, "transmit_window_errors", int64(msg.Attributes.Stats64.TXWindowErrors), msg.Attributes.Name)
+
+		// for cslip etc
+		pushStats(ch, c.subsystem, "receive_compressed", int64(msg.Attributes.Stats64.RXCompressed), msg.Attributes.Name)
+		pushStats(ch, c.subsystem, "transmit_compressed", int64(msg.Attributes.Stats64.TXCompressed), msg.Attributes.Name)
+		pushStats(ch, c.subsystem, "receive_nohandler", int64(msg.Attributes.Stats64.RXNoHandler), msg.Attributes.Name)
+
+	}
+
+	return nil
+}
+
+func pushMetricRTNL(ch chan<- prometheus.Metric, subsystem string, name string, value int64, ifaceName string, valueType prometheus.ValueType) {
+	fieldDesc := prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, subsystem, name),
+		fmt.Sprintf("%s value of /sys/class/net/<iface>.", name),
+		[]string{"device"},
+		nil,
+	)
+
+	ch <- prometheus.MustNewConstMetric(fieldDesc, valueType, float64(value), ifaceName)
+}
+
+func pushStats(ch chan<- prometheus.Metric, subsystem string, name string, value int64, ifaceName string) {
+	fieldDesc := prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, subsystem, name+"_total"),
+		fmt.Sprintf("Network device statistic %s.", name),
+		[]string{"device"},
+		nil,
+	)
+
+	ch <- prometheus.MustNewConstMetric(fieldDesc, prometheus.CounterValue, float64(value), ifaceName)
+}
+
+func (c *netClassRTNLCollector) getNetClassInfoRTNL() ([]LinkMessage, error) {
+	conn, err := netlink.Dial(unix.NETLINK_ROUTE, nil)
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+
+	req := rtnetlink.LinkMessage{}
+	flags := netlink.Request | netlink.Dump
+	reqMsg, err := c.packMessage(&req, unix.RTM_GETLINK, flags)
+	if err != nil {
+		return nil, err
+	}
+	repMsgs, err := conn.Execute(reqMsg)
+	if err != nil {
+		return nil, err
+	}
+	lMsgs, err := c.unpackMessages(repMsgs)
+
+	return lMsgs, err
+
+}
+
+func (c *netClassRTNLCollector) packMessage(m *rtnetlink.LinkMessage, family uint16, flags netlink.HeaderFlags) (netlink.Message, error) {
+
+	nm := netlink.Message{
+		Header: netlink.Header{
+			Type:  netlink.HeaderType(family),
+			Flags: flags,
+		},
+	}
+
+	mb, err := m.MarshalBinary()
+	if err != nil {
+		return netlink.Message{}, err
+	}
+	nm.Data = mb
+
+	return nm, nil
+}
+
+func (c *netClassRTNLCollector) unpackMessages(msgs []netlink.Message) ([]LinkMessage, error) {
+	lmsgs := make([]LinkMessage, 0, len(msgs))
+
+	for _, nm := range msgs {
+		var m LinkMessage
+		switch nm.Header.Type {
+		case unix.RTM_GETLINK, unix.RTM_NEWLINK, unix.RTM_DELLINK:
+
+		default:
+			level.Warn(c.logger).Log("msg", "NetClassCollector: unexpected message type", "type", nm.Header.Type)
+			continue
+		}
+
+		if err := c.unmarshalBinary(&m, nm.Data); err != nil {
+			return nil, err
+		}
+		lmsgs = append(lmsgs, m)
+	}
+
+	return lmsgs, nil
+}
+
+func (c *netClassRTNLCollector) unmarshalBinary(m *LinkMessage, b []byte) error {
+	l := len(b)
+	if l < unix.SizeofIfInfomsg {
```

Comment:
This length check will fail depending on which kernel is running on the system.
The constant in package `unix` does not take the kernel version into account.

Reply:
> As the author of the rtnetlink library I had quite some contact with the netlink library author befor. I know of this tool he created (using netlink) that might add the missing information you are looking for. Perhaps it is worth looking into.. https://github.com/mdlayher/ethtool/blob/main/client.go from glancing the code it should at least cover the cases of speed and duplex.
> 
> it uses the ethtool genetlink family.

Thank you very much for the review ! it is great to have the author of rtnetlink helping me :D 
I will try ethtool netlink interface to fetch the missing 4 metrics and update this PR. 
By the way, do you think it useful to add all the fields in the netlink message RTM_NEWLINK back to the structure rtnetlink.LinkAttributes? for example the carrier state fields.

Reply:
Which version of the kernel has a different value than 16 for `SizeofIfInfomsg` ? 👀
After a quick check in kernel code, it seems the definition of `struct ifinfomsg` stays unchanged in version [3.19](https://github.com/torvalds/linux/blob/v3.19/include/uapi/linux/rtnetlink.h#L446), [4.0](https://github.com/torvalds/linux/blob/v4.0/include/uapi/linux/rtnetlink.h#L453), [5.18](https://github.com/torvalds/linux/blob/v5.18/include/uapi/linux/rtnetlink.h#L557)

Reply:
Hmm, I might have spoken out in error. You are right, that one actually never changed. I apologize for the confusion!

Reply:
no problem